numpy/core/tests/test_multiarray.py

```diff
@@ -3548,21 +3566,114 @@ def test_test_zero_rank(self):
+                array.dump(f, protocol=5)
+
+            with pytest.raises(ImportError):
+                array.dumps(protocol=5)
+
+            with pytest.raises(ImportError):
+                array.__reduce_ex__(5)
+
+        elif sys.version_info[:2] < (3, 6):
+            # when calling __reduce_ex__ explicitly with protocol=5 on python
+            # version <3.6, instead of raising an AttributeError saying "pickle
+            # module has no attribute PickleBuffer", raise a ValueError
+            # saying that protocol 5 is not available for this python version
+            with pytest.raises(ValueError):
+                array.__reduce_ex__(5)
+
+            with pytest.raises(ValueError):
+                array.dumps(protocol=5)
+
+            with pytest.raises(ValueError):
+                array.dump(f, protocol=5)
+
+    @pytest.mark.skipif(pickle.HIGHEST_PROTOCOL < 5,
+                        reason="requires pickle protocol 5")
+    def test_record_array_with_object_dtype(self):
+        my_object = object()
+
+        arr_with_object = np.array(
+                [(my_object, 1, 2.0)],
+                dtype=[('a', object), ('b', int), ('c', float)])
+        arr_without_object = np.array(
+                [('xxx', 1, 2.0)],
+                dtype=[('a', str), ('b', int), ('c', float)])
+
+        arr_with_object_bytes_string = pickle.dumps(arr_with_object,
+                                                    protocol=5)
+        arr_without_object_bytes_string = pickle.dumps(arr_without_object,
+                                                       protocol=5)
+
+        assert b'frombuffer' in arr_without_object_bytes_string
+        assert b'frombuffer' not in arr_with_object_bytes_string
+
+    @pytest.mark.skipif(pickle.HIGHEST_PROTOCOL < 5,
+                        reason="requires pickle protocol 5")
+    def test_f_contiguous_array(self):
+        # Case of fortran-contigous array
+        f_contiguous_array = np.array([[1, 2, 3], [4, 5, 6]], order='F')
+
+        # When using pickle protocol 5, Fortran-contiguous arrays can be
+        # serialized using PickleBuffer
+        bytes_string = pickle.dumps(f_contiguous_array, protocol=5)
+        assert b'frombuffer' in bytes_string
+        depickled_f_contiguous_array = pickle.loads(bytes_string)
+        assert_equal(f_contiguous_array, depickled_f_contiguous_array)
+
+    @pytest.mark.skipif(pickle.HIGHEST_PROTOCOL < 5,
+                        reason="requires pickle protocol 5")
+    def test_non_contiguous_array(self):
+        non_contiguous_array = np.arange(12).reshape(3, 4)[:, :2]
+        assert not non_contiguous_array.flags.c_contiguous
+        assert not non_contiguous_array.flags.f_contiguous
+
+        # This array is neither c nor f-style contiguous, thus it cannot be
+        # pickled using PickleBuffers
+        bytes_string = pickle.dumps(non_contiguous_array, protocol=5)
+        assert b'frombuffer' not in bytes_string
```

Comment:
Here again I don't see the use of testing that it's not possible to do an optimization via an implementation detail.

This test it still useful to show that pickling is not broken on non-contiguous arrays but I would remove the assertion on the frombuffer implementation detail, change the comment to say that this is a test to ensure that non-contiguous arrays can be pickled with any protocol (although the no-copy semantics is no longer guaranteed). Remove the `@pytest.mark.skipif`